### ISSUE TITLE
Bug: `ExtractDispatchActions<unknown>` inferred from multiple models

### ISSUE THREAD

**Issue by hackape** [modern-js-dev/reduck#30]
When pass in multiple models to `useModel`, if the `"actions"` field is missing in any model, the final merged `actions` is inferred as `ExtractDispatchActions<unknown>`. If supplied an empty object as `"actions"` value, the type is inferred correctly.

Reproduction:

```ts
const fooModel = model("foo").define({
  state: {
    foo: "foo",
  },
  // un-comment this line to obtain correct inferred type.
  // actions: {}
})

const countModel = model("counter").define({
  state: {
    value: 1,
  },
  actions: {
    add(state) {
      return { ...state, value: state.value + 1 }
    }
  }
})

const [state, actions] = store.use(fooModel, countModel);
             // ^ ts inferred type: const actions: ExtractDispatchActions<unknown>
```

**Comment by hackape** [comment]
https://github.com/modern-js-dev/reduck/blob/6381efec31d0307cc6d63e9a4ca52a9ea20f4c56/packages/store/src/model/model.ts#L72-L74

I'd like to suggest a quick patch here.

```diff
 response._ = undefined as Omit<M, 'state'> & { 
   state: State extends void ? M['state'] : State; 
+  actions: 'actions' extends keyof M ? M['actions'] : {};
 }; 
```

**Comment by anc95** [comment]
> Why prefer model(name).define(modelDesc) over model({ name, ...modelDesc })

At first I wanted to implement it in this way `model({ name, ...modelDesc })`, But the following TS type of problem was encountered
```ts
const todo = model<State>({
  name: 'a',
  state: { value: 1 }
  actions: {
    add(state) {
      state.value++;
    }
  }
});

store.use(todo);
```
```ts
type model<State> = (modelDesc: ModelDesc<State>) => Model<State>
```

Here the `Model` cannot be derived from the real `ModelDesc` passed in by the user

```ts
type model<State> = (name: string) => { define: <M extends ModelDesc<M>>(modelDesc: M) => Model<M> }
```

In this way, it is convenient to constrain the type of `modelDesc` by the generic type of the `model function`, and the final returned `Model` can contain the real type information of `modelDesc`, so that it is convenient to `useModel` to derive `actions` and `state` type.

To summarize, the use of `model(name).define(modelDesc)` is actually a trade-off to better support TS

**Comment by anc95** [comment]
> ```diff
> +  actions: 'actions' extends keyof M ? M['actions'] : {};
> ```

I think it does fix what issue you report, cloud can raise a PR?, while I also found some other problem of type derive, I would fix it myself

**Comment by anc95** [comment]
> > ```diff
> > +  actions: 'actions' extends keyof M ? M['actions'] : {};
> > ```
> 
> I think it does fix what issue you report, cloud can raise a PR?, while I also found some other problem of type derive, I would fix it myself

I have solved this problem in #31

**Comment by hackape** [comment]
> "TS type problem was encountered"

I don't understand what exactly is the problem encountered? It's not obvious from your sample code.

I construct the following `model` function interface to showcase my preferred signature. It works in both cases, as seen below. You can either provide a `State` type argument to constrain `ModelDesc`, or you can omit it and fallback to infer the actual state type from `ModelDesc`.

```ts
type Model<S> = {
    (name: string): any;
    _: Omit<ModelDesc<S>, 'name'>;
    _name?: string;
};

interface ModelDesc<State = any> {
    name: string;
    state: State;
    actions?: Actions<State>;
}

type Actions<S> = {
    [actionType: string]: (state: S, ...args: any) => any
};

function model<S>(modelDesc: ModelDesc<S> | ((ctx: any, hooks: any) => ModelDesc<S>)): Model<S> {
    return undefined as any
}

// usage
type State = { value: number };
const todo1 = model<State>({
    name: 'a',
    state: { value: 1 },
    actions: {
        add(state) {
            state.value++;
        }
    }
});

const todo2 = model({
    name: 'a',
    state: { value: 1 },
    actions: {
        add(state) {
            state.value++;
        }
    }
});

```

[Playground][1]


[1]: https://www.typescriptlang.org/play?#code/C4TwDgpgBAsg9gEwgGwDwGUB8UC8UDeAUFCVABQB2AhgLYQBcUAzsAE4CWFA5gJSNUUQAbmKkA+owDyNdsFTwkyACIQmAYwyYANFADk1OrswjSUMQYgB+Riw7cRAXxGFOwCKwBmVNdAUoV6hjAVG64UAIg2ESmFjZsnFwmpCwhDFDowW5JJN7A7HAUTNZQAIJqeQVMQanGhA6EhKCQpeX5hZph0aQA2rltACrgabYJALqMZClujOg6AHQLVKxcTPyCPLjYEXXOHgCuFK0FUDSIKJpkp4oBaox+yqoaWFAAPuRk5QAeayA6ABZwOAAa1W4XWm1gZwegSwPD4kMUHS6JFYEGAe1YFCgByQHk4EAQ4SYYJAdQaAHpydimFQuBBGkN0ploHh8FAAG5UZB7NIUPY0ABG7igTkIakqwCgwEQcAAjGErucMjUyMioLE9FRdFpRCQpmk2ZzuWl5Q4daY+pVGGqLQgEJNmRsbaZksy5kaeQBqT3ZF31Uz1Bw8ZziwqS6UIOAAJgVUNVuvVtDSui15tdqWtHK5PMYprTOSOhWtCdt9v1TpLLr1bo9EG9voDCcDwcIQA

**Comment by anc95** [comment]
> Playground

@hackape You can try to implement a simple type of useModel based on the above code

**Comment by hackape** [comment]
OK, here's my experiment.

I copy-paste everything from [/packages/store/src/types/model.ts](https://github.com/modern-js-dev/reduck/blob/6381efec31d0307cc6d63e9a4ca52a9ea20f4c56/packages/store/src/types/model.ts), and apply only one modification to `type Model` to add a generic type param `<S>` for `State`. Everything else stays the same. 

```ts
type Model<S=any> = {
  (name: string): any;
  _: Omit<ModelDesc<S>, 'name'>;
  // model name
  _name?: string;
};
```

Then I use `function model<S>(modelDesc: ModelDesc<S> | ((ctx: any, hooks: any) => ModelDesc<S>)): Model<S>` to create 2 models, just like in earlier code snippet. Turns out they work just fine with current `useModel` function interface.

I don't encounter any problem. 😕 So...I still don't get what exact is the TS problem that you mentioned. Did I miss anything?

[Playground link][1].

![image](https://user-images.githubusercontent.com/8945315/141952956-0cf25134-5437-4916-b79d-de2dd098ba1f.png)

[1]: https://www.typescriptlang.org/play?strictNullChecks=false#code/C4TwDgpgBAwg9gO2BAHsKBeKBDBIDcAUIamHAE7qiRQCCAxsAJaIA8AysNsgDRQAK2EABs42ACaYceAHxSAFIShQAzl2QAuKJ24QeSqGCGiJWwSLHj9ygHR3UwctlrkA5iq24QAbQC6hAEpMOQB6EKgIFWEmJABacSYVbAAjYQhYhAdY6MyoAAFqSPpyJjBgWMic4BCEOFiYgDdsaPFYhrgmVsLCHWQoAB8ods6iEhQySigY5HIAM2x6aAZmRBUOdQg5AG8DbwBrCBAtNRKEV18tZZYEdd05QavV2+QZIgBfYjCIqOn4xJS0hksjloAVwEUSmUKj8kDU6hAALZlED1JAQOYLCBjCboabo+aLKAAWTg4ggwgAIkUAPJla4qKBbD6EL6VX4JJKpdKZNDZGKgwoqYqlcps2G1WIAVwQkpUEFaTXIKmxFFxaIxhJJZMpRWe0CwXm2BhCACoTQYoCaoABVBBMACOkugCGwCKxyktIQMLrdx0cMVcRGUpvNHqtWvJqg2FpNXuUal0Wl6ECDUBDMeJpMjCxWCBUfCYwAA5AzsIZJeRoLNpYxrlAAO6FgAWUFqEVmswgjBjcZwtdWAH5Lv283rXoRmaRVVN1QToBTEkZgPQm4884zdgcjlHTuctPJ5EYLKZpCAghg5MNxEFBguVEuV2uVO9iFPJoVM9qOBhDVIdsp5B9CA-V3AJPDwVMAH0tGpBFC1YCMdSFDgZD4IsgKLcdg3CBEs2EVtXXdKBIKAocdwDd5Rg-Il0VcCBWAAFQiNAIAQcQGQAJS7ChxFYE4Az4Q0+AAaWY5A2IZfizjkLB-ygfZDhnKAtzgWYoCY7AGWEi51IUkBfEo4gP1ta4GLgABJdU5RHVhrRkqBFGUa0xNY9jTygAcHOwNwtGtc85C8KAtEyBp0UCFyJK8nyZ07cgoHM-yhg6cQDE88yDGCiBQvIUZ00tKAAHEIGAJ8lCtWgoA-Js4DgPZKrgKBKxUOBhFCvtcwZD9ZnIOAEWURCyq9N81RmOdCuKp8EIitzEKNZQc3pLQAFE0CcRg7wfVcRzWIlvCLSCi18PaFtWQ7x2ZPKrSK4BkzKuhKvBKBqtq+rGsiFq2oTPoup6vrP3JQaVUmPENWga7kymhxXIZWaN3jDYtF2-bDr2r6IEOl8gaoR6SWlZBxEQyGWMixCpFh2TvUIkCKIMNHfLtRAzMsmZrNzBDaPo8GNgQ1CoBLDZMNRgXfCw9rFptBmECZqyuzZmi3E5ibtp5tCTrzQWizVlQztTXDtSpIVEbwg36EMlkzTulbHBzMXECgDlNvRKBut65Q9ezbb7YgWZ+RiuA7ryVBXTANIyru5JvYod0rX-K15u2rRY49eb9zp7QNj4I8THEamzkSpk7uUD58uL0MrWwWYZjupP446xPC49bBU4R9PdHzt4gzjqBS+7wGPyttbgA27hH2ViqociybfywKfZHsuS9KUlS1NoHTaD03xpoZWeQBkC0PKgAec2H5cto61h163EX98uDet4cxKvH3zz5ESq8b7oO+J7c+R2HAkA+B2BsN5dwWgYixW0I-PAz8HJAJAR4SBwQkqdA-iFdEZtqJ4QYo9LAiEqKPWuohFQEMiH30Qn4eyJlGYWRljZAwuM0QEzwghPCKhvAygRBHcgIsha6EOoQcchAPyENYWuFh2oGTfxhnhChUgqFSxoSzWW1xWD0LgHjeUhMiHsMlJw9EPDNbbX4YI4RxUiHJgYpKEO9E1ESLIVg8EfhrA2nvl4PwUgnECLJqw++3hwFOwABIQAkIAuw-i4oMWwEwYQ-hlCeSCRIex2psGQBgZE6JSTyQpIgH4fecTxrAHMRsSx1jGJROEHwbwQDrR8BERIiG3gEniBFtfZOygtBVLsDUgpRTdCsEacE5pMhYkel8hlG0+CaB1PJCoJ8JS0iqP6j4qR-1hDZM8U5VxeB3FYE8fZUhKy-EIAgU00JNhwnqXKSM+JgzMlrPBGk8pdz1kjOTp5aZwhZnbXmfRdJFT5LVNqWY0RNkBkSBaXvNpQUAVdKBYUkFbMwVDNedC604zrSjGGjOUamIbRylJnJL4soICIXkNgRCfBkiIQCKmcRMy7kUPkEA92nyjYSLAvJfeHziHcyIahLlwKJFiL5c45QXFgAViluCVg8A0RoGOmARIKMix0XhdqdgkpkhChKBHM6ziDLGnCMS0l3haAUqgAAIUQr4GlBg6WfIZSLeQLKEGdJsEQm1HSBVqpmSQ1h-Lk7cuFf60VUBxWSuyTKxAyB5Wa0VdrI6KrBXkg1Vq4UuqRb6tTES-FeF5C0BZXwK1eEzl8DRuwckssKCCW2hWtIjAKABDtfvA5xMZoyN8KG5Mdaq1xRWUyuw8CtDcosVYhZfKoEAP3muHtDa+1toZAO4BbgEFBu+WO+iE6kFeGcTIRyHo4HRTdS6tlMyy1kW7ZWudgkyIzqvcACgIyOXeH3pe+tD753iR-oe0Bp4-CJQuewA+QHMrZVDXe99FB75LqHX+m1SDAPAehWg8gobw3kClZAKNcrgAKqVYm1ViFU3aqYBmgNUADXYSgMavNBaS2WvNUAstGxZ0fqbcoRZHpW1fukdqDZHo329ugz+1dybPmjtKVui8p5d37tsHYE9ML3WsL4IJudT6vXJzUx+4Tg6V3-3-Qh45TsgOeRA62LK6JQ1rrliG-e6HMP0VlTG3Dcb8N7UI3hYj6b0aZoMJRtMRrc3ankKawtDH6NAM7VGXQrHq22zzHF8g7GoCcaWXYlZ5DO2vpY-eqD-aRPDrEzy3QPz7UqBkJO8DI4ku6eXb+mz9IytSYCngWT+9FNuo9aGlQF7cuQdQ-vbAt6at5cG8oZ9OXYtjbq7Btx8HpOIbM8hyz42PQQaEwVvTv75sAeM3FUz2gVtgfs8VCN0rnMODwwmjzYnvM6t8+RgLOaSV5rC-R4t2oznRfLWNlLaXVmSIXas-jyhtP5eBzB-TPTWESfHf6qrBg90ddYR0oB3X969aTP13tzjJtaZx3O2b0PdtGYgYd8zKHrPFeDRI8jYqzsYcjZd2N2B43Ks8+qzVJGyNZonMQAXl1iTUgpNoBitAOIMSAwEpaHElqA2INWBAI4oAspQs642RRT1IXoChAYDl5CMBQP-Pgz09gIK8IlRCJsUIBA5YTdgc0PSVglRh6jbFvb8kkJpU8-OWRBewHRIRj1kx-iGM0J0wVdFcO7kQegqwqCkjgAARikGr5Me65LKCAloTWRYesty2OH4QkeoCp7eKGrW9coXzXEOIeQaMghZ5r-DXQNgmgl4gAAai76mKFxcPQfDeLawg8e8yJ-EHAAATGnvNzec982wPnzHhfnY1S0OXyvCc4Y1+SLosADeNhN7yTXtGNhZg1R733tpA+i4ThH2kdANHtS+WC+SUYxAx9qHkmjGtHVN5YDP7kjyAPqT7J58CgHT58BkqUqJTyBbBwI8BALJDD5NpAA

**Comment by hackape** [comment]
To be crystal clear, I'm not hard-selling anything here. I understand that even if I had a good point, the API is likely to stay since you've already shipped v1 and breaking change is unlikely to happen.

I had this confusion reading the source code, and here's me sharing my thoughts and that's all. Like I said it's a side question /discussion, not a problem that needs fixing.

**Comment by anc95** [comment]
> [Playground link](https://www.typescriptlang.org/play?strictNullChecks=false#code/C4TwDgpgBAwg9gO2BAHsKBeKBDBIDcAUIamHAE7qiRQCCAxsAJaIA8AysNsgDRQAK2EABs42ACaYceAHxSAFIShQAzl2QAuKJ24QeSqGCGiJWwSLHj9ygHR3UwctlrkA5iq24QAbQC6hAEpMOQB6EKgIFWEmJABacSYVbAAjYQhYhAdY6MyoAAFqSPpyJjBgWMic4BCEOFiYgDdsaPFYhrgmVsLCHWQoAB8ods6iEhQySigY5HIAM2x6aAZmRBUOdQg5AG8DbwBrCBAtNRKEV18tZZYEdd05QavV2+QZIgBfYjCIqOn4xJS0hksjloAVwEUSmUKj8kDU6hAALZlED1JAQOYLCBjCboabo+aLKAAWTg4ggwgAIkUAPJla4qKBbD6EL6VX4JJKpdKZNDZGKgwoqYqlcps2G1WIAVwQkpUEFaTXIKmxFFxaIxhJJZMpRWe0CwXm2BhCACoTQYoCaoABVBBMACOkugCGwCKxyktIQMLrdx0cMVcRGUpvNHqtWvJqg2FpNXuUal0Wl6ECDUBDMeJpMjCxWCBUfCYwAA5AzsIZJeRoLNpYxrlAAO6FgAWUFqEVmswgjBjcZwtdWAH5Lv283rXoRmaRVVN1QToBTEkZgPQm4884zdgcjlHTuctPJ5EYLKZpCAghg5MNxEFBguVEuV2uVO9iFPJoVM9qOBhDVIdsp5B9CA-V3AJPDwVMAH0tGpBFC1YCMdSFDgZD4IsgKLcdg3CBEs2EVtXXdKBIKAocdwDd5Rg-Il0VcCBWAAFQiNAIAQcQGQAJS7ChxFYE4Az4Q0+AAaWY5A2IZfizjkLB-ygfZDhnKAtzgWYoCY7AGWEi51IUkBfEo4gP1ta4GLgABJdU5RHVhrRkqBFGUa0xNY9jTygAcHOwNwtGtc85C8KAtEyBp0UCFyJK8nyZ07cgoHM-yhg6cQDE88yDGCiBQvIUZ00tKAAHEIGAJ8lCtWgoA-Js4DgPZKrgKBKxUOBhFCvtcwZD9ZnIOAEWURCyq9N81RmOdCuKp8EIitzEKNZQc3pLQAFE0CcRg7wfVcRzWIlvCLSCi18PaFtWQ7x2ZPKrSK4BkzKuhKvBKBqtq+rGsiFq2oTPoup6vrP3JQaVUmPENWga7kymhxXIZWaN3jDYtF2-bDr2r6IEOl8gaoR6SWlZBxEQyGWMixCpFh2TvUIkCKIMNHfLtRAzMsmZrNzBDaPo8GNgQ1CoBLDZMNRgXfCw9rFptBmECZqyuzZmi3E5ibtp5tCTrzQWizVlQztTXDtSpIVEbwg36EMlkzTulbHBzMXECgDlNvRKBut65Q9ezbb7YgWZ+RiuA7ryVBXTANIyru5JvYod0rX-K15u2rRY49eb9zp7QNj4I8THEamzkSpk7uUD58uL0MrWwWYZjupP446xPC49bBU4R9PdHzt4gzjqBS+7wGPyttbgA27hH2ViqociybfywKfZHsuS9KUlS1NoHTaD03xpoZWeQBkC0PKgAec2H5cto61h163EX98uDet4cxKvH3zz5ESq8b7oO+J7c+R2HAkA+B2BsN5dwWgYixW0I-PAz8HJAJAR4SBwQkqdA-iFdEZtqJ4QYo9LAiEqKPWuohFQEMiH30Qn4eyJlGYWRljZAwuM0QEzwghPCKhvAygRBHcgIsha6EOoQcchAPyENYWuFh2oGTfxhnhChUgqFSxoSzWW1xWD0LgHjeUhMiHsMlJw9EPDNbbX4YI4RxUiHJgYpKEO9E1ESLIVg8EfhrA2nvl4PwUgnECLJqw++3hwFOwABIQAkIAuw-i4oMWwEwYQ-hlCeSCRIex2psGQBgZE6JSTyQpIgH4fecTxrAHMRsSx1jGJROEHwbwQDrR8BERIiG3gEniBFtfZOygtBVLsDUgpRTdCsEacE5pMhYkel8hlG0+CaB1PJCoJ8JS0iqP6j4qR-1hDZM8U5VxeB3FYE8fZUhKy-EIAgU00JNhwnqXKSM+JgzMlrPBGk8pdz1kjOTp5aZwhZnbXmfRdJFT5LVNqWY0RNkBkSBaXvNpQUAVdKBYUkFbMwVDNedC604zrSjGGjOUamIbRylJnJL4soICIXkNgRCfBkiIQCKmcRMy7kUPkEA92nyjYSLAvJfeHziHcyIahLlwKJFiL5c45QXFgAViluCVg8A0RoGOmARIKMix0XhdqdgkpkhChKBHM6ziDLGnCMS0l3haAUqgAAIUQr4GlBg6WfIZSLeQLKEGdJsEQm1HSBVqpmSQ1h-Lk7cuFf60VUBxWSuyTKxAyB5Wa0VdrI6KrBXkg1Vq4UuqRb6tTES-FeF5C0BZXwK1eEzl8DRuwckssKCCW2hWtIjAKABDtfvA5xMZoyN8KG5Mdaq1xRWUyuw8CtDcosVYhZfKoEAP3muHtDa+1toZAO4BbgEFBu+WO+iE6kFeGcTIRyHo4HRTdS6tlMyy1kW7ZWudgkyIzqvcACgIyOXeH3pe+tD753iR-oe0Bp4-CJQuewA+QHMrZVDXe99FB75LqHX+m1SDAPAehWg8gobw3kClZAKNcrgAKqVYm1ViFU3aqYBmgNUADXYSgMavNBaS2WvNUAstGxZ0fqbcoRZHpW1fukdqDZHo329ugz+1dybPmjtKVui8p5d37tsHYE9ML3WsL4IJudT6vXJzUx+4Tg6V3-3-Qh45TsgOeRA62LK6JQ1rrliG-e6HMP0VlTG3Dcb8N7UI3hYj6b0aZoMJRtMRrc3ankKawtDH6NAM7VGXQrHq22zzHF8g7GoCcaWXYlZ5DO2vpY-eqD-aRPDrEzy3QPz7UqBkJO8DI4ku6eXb+mz9IytSYCngWT+9FNuo9aGlQF7cuQdQ-vbAt6at5cG8oZ9OXYtjbq7Btx8HpOIbM8hyz42PQQaEwVvTv75sAeM3FUz2gVtgfs8VCN0rnMODwwmjzYnvM6t8+RgLOaSV5rC-R4t2oznRfLWNlLaXVmSIXas-jyhtP5eBzB-TPTWESfHf6qrBg90ddYR0oB3X969aTP13tzjJtaZx3O2b0PdtGYgYd8zKHrPFeDRI8jYqzsYcjZd2N2B43Ks8+qzVJGyNZonMQAXl1iTUgpNoBitAOIMSAwEpaHElqA2INWBAI4oAspQs642RRT1IXoChAYDl5CMBQP-Pgz09gIK8IlRCJsUIBA5YTdgc0PSVglRh6jbFvb8kkJpU8-OWRBewHRIRj1kx-iGM0J0wVdFcO7kQegqwqCkjgAARikGr5Me65LKCAloTWRYesty2OH4QkeoCp7eKGrW9coXzXEOIeQaMghZ5r-DXQNgmgl4gAAai76mKFxcPQfDeLawg8e8yJ-EHAAATGnvNzec982wPnzHhfnY1S0OXyvCc4Y1+SLosADeNhN7yTXtGNhZg1R733tpA+i4ThH2kdANHtS+WC+SUYxAx9qHkmjGtHVN5YDP7kjyAPqT7J58CgHT58BkqUqJTyBbBwI8BALJDD5NpAA)

I've seen that the implementation here, it seems not working well on deriving the type of actions. My previous implementation was similar to yours, but the automatic derivation of the `actions` type when passing in the generic `State` bothered me for a long time, so I settled for `model<State>(name).define({})` instead.

**Comment by anc95** [comment]
> So...I still don't get what exact is the TS problem that you mentioned. Did I miss anything?

Maybe the language is difficult to describe, I'll put together a TS demo later

**Comment by hackape** [comment]
Sure, when you got the time. And if you don’t then just leave it, it’s not that important anyway. I kinda feel sorry for holding yon up for too long just to satisfy my curiosity 😅
